### ISSUE TITLE
Handling NSErrors returned from factory methods that produce errors.

### ISSUE THREAD

**Issue by fatuhoku** [appsquickly/Typhoon#326]
By using `-withFactory:selector:` it's possible to invoke factory methods that throw errors. e.g.

```objc
- (MESFooBar *)foobar {
    return [TyphoonDefinition withFactory:[self foobarFactory] selector:@selector(foobarWithError:)]; // ...
}
```

The question is, how can the error be processed if it arises?

**Comment by alexgarbarev** [comment]
First of all, use HEAD version of Typhoon. (Recently I did minor change to get this feature working)
Then change your assembly like that:
```
- (MESFooBar *)foobarWithError:(NSValue *)error {
    return [TyphoonDefinition withFactory:[self foobarFactory] selector:@selector(foobarWithError:) parameters:^(TyphoonMethod *factoryMethod) {
               [factoryMethod injectParameterWith:error];
           }]]; 
}
```

**Option 1**
Method declaration:
```
- (MESFooBar *)foobarWithError:(NSError **)error
```
Caller:
```
__autoreleasing NSError *error = nil;
MESFooBar *foobar = [self.assembly foobarWithError:[NSValue valueWithPointer:&error]];
if (error) {
    // error handling
}
````

**Option 2**

Method declaration:
```
- (MESFooBar *)foobarWithError:(NSError *__strong*)error
```
Caller:
```
NSError *error = nil;
MESFooBar *foobar = [self.assembly foobarWithError:[NSValue valueWithPointer:&error]];
if (error) {
    // error handling
}
````
**IMPORTANT NOTICE**
You have to be careful with ownership rules. ARC can't handle this case implicit, because it's too complex for ARC (object written by reference wrapped by NSValue). So you have to use explicit ownership rules:
* if NSError on caller side is strong (have __strong attribute or empty), then factory method declaration must be like (NSError \*__strong\*)
* if NSError on caller side is autoreleasing (have __autoreleasing attribute), then factory method declaration must be like (NSError \*__autoreleasing\*) or (NSError **)

Otherwise you'll get the crash

**Comment by jasperblues** [comment]
@alexgarbarev Looks like tests are failing since last commits.

**Comment by alexgarbarev** [comment]
@jasperblues, yes, reverted back. This failure related to #325 - not this ticket

**Comment by alexgarbarev** [comment]
I think we should update documentation for that common case (just faced with that and had to search for that ticket)
